services/src/main/java/org/exoplatform/officeonline/WOPIService.java

```diff
@@ -198,6 +209,93 @@ public WOPIService(SessionProviderService sessionProviders,
     return map;
   }
 
+  /**
+   * Verify proof key.
+   *
+   * @param proofKeyHeader the proof key header
+   * @param oldProofKeyHeader the old proof key header
+   * @param url the url
+   * @param accessToken the access token
+   * @param timestampHeader the timestamp header
+   * @return true, if successful
+   */
+  public boolean verifyProofKey(String proofKeyHeader,
+                                String oldProofKeyHeader,
+                                String url,
+                                String accessToken,
+                                String timestampHeader) {
+    if (StringUtils.isBlank(proofKeyHeader)) {
+      return true; // assume valid
+    }
+
+    long timestamp = Long.parseLong(timestampHeader);
+    if (!ProofKeyHelper.verifyTimestamp(timestamp)) {
+      return false;
+    }
+
+    byte[] expectedProofBytes = ProofKeyHelper.getExpectedProofBytes(url, accessToken, timestamp);
+    // follow flow from https://wopi.readthedocs.io/en/latest/scenarios/proofkeys.html#verifying-the-proof-keys
+    boolean res = ProofKeyHelper.verifyProofKey(discoveryPlugin.getProofKey(), proofKeyHeader, expectedProofBytes);
+    if (!res && StringUtils.isNotBlank(oldProofKeyHeader)) {
+      res = ProofKeyHelper.verifyProofKey(discoveryPlugin.getProofKey(), oldProofKeyHeader, expectedProofBytes);
+      if (!res) {
+        res = ProofKeyHelper.verifyProofKey(discoveryPlugin.getOldProofKey(), proofKeyHeader, expectedProofBytes);
+      }
+    }
+    return res;
+  }
+
+  /**
+   * Sets the plugin.
+   *
+   * @param plugin the plugin
+   */
+  public void setWOPIDiscoveryPlugin(ComponentPlugin plugin) {
+    Class<WOPIDiscoveryPlugin> pclass = WOPIDiscoveryPlugin.class;
+    if (pclass.isAssignableFrom(plugin.getClass())) {
+      discoveryPlugin = pclass.cast(plugin);
+      LOG.info("Set WopiDiscoveryPlugin instance of " + plugin.getClass().getName());
+    } else {
+      throw new WopiDiscoveryNotFoundException("WopiDiscoveryPlugin is not an instance of " + pclass.getName());
+    }
+  }
+
+  @Override
+  public void start() {
+    if (discoveryPlugin == null) {
+      throw new WopiDiscoveryNotFoundException("WopiDiscoveryPlugin is not configured");
+    }
+    discoveryPlugin.start();
+
+    if (activeCache.get(SECRET_KEY) == null) {
+      activeCache.put(SECRET_KEY, generateSecretKey());
```

Comment:
We should be sure that before this line we don't need a secret key, otherwise it should be generated in the constructor - configured or generated, but in a single place is better.

Reply:
Fixed in https://github.com/exo-addons/office-online/pull/2/commits/1a2933bb7099ed744368596f3d724b7396b29bd2